### PR TITLE
[WIP] Static cluster tls

### PR DESCRIPTION
**Pull request by colhom**
\cc @hongchaodeng @xiang90 

Still testing this out, will ping once it's passing the tire kicking tests.

### CODE REVIEW

File: example/example-etcd-cluster.yaml [hunk 1]
```diff
@@ -5,3 +5,7 @@ metadata:
 spec:
   size: 3
   version: "3.1.0"
+  clusterTLS:
```
Comment (xiang90) [inline example/example-etcd-cluster.yaml]: TLS
  - Static
    - caSecret
    - serverSecret
    - clientSecret

Comment (xiang90) [inline example/example-etcd-cluster.yaml]: is this better?

File: glide.lock [hunk 2]
```diff
@@ -420,4 +419,49 @@ imports:
   - 1.5/tools/metrics
   - 1.5/tools/record
   - 1.5/transport
+- name: k8s.io/kubernetes
```
Comment (xiang90) [inline glide.lock]: use client-go instead? we do not want to introduce the entire k8s as dependency.

File: example/example-etcd-cluster.yaml (hunk 1 above)
Comment (colhom) [inline example/example-etcd-cluster.yaml]: Definitely, pushed a commit that scopes these fields. I did leave the `Name` suffix on the field names, to indicate it's a pointer to a secret and not a literal.

File: glide.lock (hunk 2 above)
Comment (colhom) [inline glide.lock]: yep- done.

**Comment by colhom** [comment]
I realize now that there are a bunch of `http://` --> `https://` references I haven't switched over yet, which probably is not helping this thing pass e2e tests. Stand by

File: pkg/util/tlsutil/x509.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+	"time"
+)
+
+var (
+	Duration365d = time.Hour * 24 * 365
```
Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: 24 * 365 .* time.Hour

File: pkg/util/tlsutil/x509.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+//Certificate Signers (things which mint identities. Pulling the interface spec out now will make implementing new ones cake)
+
+// This is explictly not a certificate request model
+// ie: the singee must divulge private key material to singer
+// This is fine for a one-way trust delegation model (CA downwards)
+type CertificateSigner func(pubKey crypto.PublicKey, certTmpl *x509.Certificate) (*x509.Certificate, error)
+
+// We have the ca private key in memory already
```
Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: NewLocalCerticateSinger return a CertificateSinger that...

Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: 'We have the ca private key in memory already' why do we need this comment?

File: pkg/util/tlsutil/x509.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+		cert, err := x509.ParseCertificate(certDERBytes)
+		if err != nil {
+			panic(fmt.Sprintf("cannot parse x509 certificate that was just generated. this should never happen %v", err))
+		}
+
+		return cert, nil
+	}
+}
+
+//Identity Factories
```
Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: remove this comment?

File: pkg/util/tlsutil/x509.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+func NewSignedServerIdentity(cfg CertConfig, signer CertificateSigner) (*x509.Certificate, error) {
+	cfg.extKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+	return newSignedIdentity(cfg, signer)
+}
+
+//TODO(chom): implement externally-signed CA
+func NewCAIdentity(cfg CACertConfig, signer CertificateSigner) (*x509.Certificate, error) {
+
```
Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: remove empty line here?

File: pkg/util/tlsutil/x509.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	if cfg.Duration <= 0 {
+		return nil, errors.New("Self-signed CA cert duration must not be negative or zero.")
```
Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: self-signed CA cert ... or zero.

File: pkg/util/tlsutil/x509.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	caCert, err := signer(cfg.PublicKey, &tmpl)
+	if err != nil {
+		return nil, fmt.Errorf("error generating self-signed CA certificate: %v", err)
+	}
+
+	return caCert, nil
+}
+
+func newSignedIdentity(cfg CertConfig, signer CertificateSigner) (*x509.Certificate, error) {
+
```
Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: remove empty line here?

File: pkg/util/tlsutil/x509.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+	serial, err := rand.Int(rand.Reader, new(big.Int).SetInt64(math.MaxInt64))
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.Duration <= 0 {
+		return nil, errors.New("Signed cert duration must not be negative or zero.")
```
Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: signed cert duration must ...

File: pkg/cluster/cluster.go [hunk 10]
```diff
@@ -87,6 +88,8 @@ type Cluster struct {
 	backupDir string
 
 	gc *garbagecollection.GC
+
+	etcdHttpClient *http.Client
```
Comment (xiang90) [inline pkg/cluster/cluster.go]: this client is used to talk to the backup sidecar, right? can we rename it to backupHTTPClient?

File: pkg/cluster/cluster.go [hunk 11]
```diff
@@ -259,6 +262,22 @@ func (c *Cluster) run(stopC <-chan struct{}) {
 				// TODO: we can't handle another upgrade while an upgrade is in progress
 				c.logger.Infof("spec update: from: %v to: %v", c.cluster.Spec, event.cluster.Spec)
 				c.cluster = event.cluster
+				staticSpec := c.cluster.Spec.ClusterTLS.Static
```
Comment (xiang90) [inline pkg/cluster/cluster.go]: let's only care TLS for the cluster creation phase initially? ignore the modification for now?

File: pkg/cluster/reconcile.go [hunk 12]
```diff
@@ -251,10 +247,7 @@ func RequestBackupNow(addr string) error {
 	return nil
 }
 
-func checkBackupExist(addr, ver string) (bool, error) {
-	httpcli := http.Client{
-		Timeout: 30 * time.Second,
-	}
+func checkBackupExist(addr, ver string, httpcli *http.Client) (bool, error) {
```
Comment (xiang90) [inline pkg/cluster/reconcile.go]: remove the empty line 251?

File: pkg/util/etcdutil/member.go [hunk 13]
```diff
@@ -38,15 +38,15 @@ func (m *Member) ClientAddr() string {
 		return strings.Join(m.ClientURLs, ",")
 	}
 
-	return fmt.Sprintf("http://%s:2379", m.Name)
+	return fmt.Sprintf("https://%s:2379", m.Name)
```
Comment (xiang90) [inline pkg/util/etcdutil/member.go]: https only if TLS is set?

File: pkg/spec/cluster_tls.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+	Static *StaticTLS `json:"static"`
+	//TODO: add more TLS options here
+}
+
+type StaticTLS struct {
+	// Secret containing peer and client CA cert files
+	CASecretName string `json:"caSecretName"`
+	// Secret containing peer-interface and client-interface server x509 key/cert
+	// Using this approach means all nodes have the same "identity", which is less than ideal
+	NodeSecretName string `json:"nodeSecretName"`
```
Comment (xiang90) [inline pkg/spec/cluster_tls.go]: can we rename this to ServerSecretName?

File: pkg/spec/cluster_tls.go [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+	// Secret containing etcd client key/cert
+	ClientSecretName string `json:"clientSecretName"`
```
Comment (xiang90) [inline pkg/spec/cluster_tls.go]: Probably we should explain why starting an etcd server needs client Secret. Like server needs to act as client to add itself into the cluster.

**Comment by xiang90** [comment]
The direction looks good. Need to cleanup a few minor issues.

File: pkg/util/tlsutil/x509.go (hunk 4 above)
Comment (colhom) [inline pkg/util/tlsutil/x509.go]: in many situations, the CA private key lives on dedicated hardware, not accessible directly by the go runtime. locking the CA keys up in a yubikey (or equivalent) is becoming more common by the day. any abstractions we have need to be compatible with that pattern.

Comment (xiang90) [inline pkg/util/tlsutil/x509.go]: The `Local` bit already explains that, right? If we want to make it more clear, can we make this comment part of the func comment of `NewLocalCertificateSigner `?

File: cmd/operator-gen-selfsigned-tls/main.go [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+	peerCertDuration, clientCertDuration time.Duration
+	prettyPrint                          bool
+	outputDir                            string
+)
+
+func init() {
+	flag.StringVar(&clusterName, "cluster-name", "default-cluster", "name of etcd operator cluster to generate tls certs for")
+	flag.StringVar(&clusterNamespace, "cluster-namespace", "default", "namespace of etcd operator cluster is running in")
+
+	flag.StringVar(&caSecretName, "ca-secret-name", "etcd-operator-ca", "name of the generated self-signed CA secret")
```
Comment (xiang90) [inline cmd/operator-gen-selfsigned-tls/main.go]: we can generate the secret name for the user based on the cluster-name, right? too many flags I feel.

File: cmd/operator-gen-selfsigned-tls/main.go [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+	flag.StringVar(&nodeClientServerCfg.CommonName, "client-node-server-cn", "etcd-node-client", "common name of self-signed etcd client server certificate")
+	flag.StringVar(&nodeClientClientCfg.CommonName, "client-node-client-cn", "etcd-client", "common name of self-signed etcd client (interface) client certificate. (used for accessing client interface)")
+
+	flag.DurationVar(&peerCACfg.Duration, "peer-ca-duration", defaultCADuration, "duration of self-signed peer certificate authority")
+	flag.DurationVar(&clientCACfg.Duration, "client-ca-duration", defaultCADuration, "duration of self-signed client certificate authority")
+
+	flag.DurationVar(&peerCertDuration, "peer-cert-duration", defaultCertDuration, "duration of self-signed peer certificates")
+	flag.DurationVar(&clientCertDuration, "client-cert-duration", defaultCertDuration, "duration of self-signed client certificates")
+
+	flag.StringVar(&outputDir, "output-directory", "./etcd-operator-selfsigned-tls", "path to output directory for tls files and kubernetes secrets")
```
Comment (xiang90) [inline cmd/operator-gen-selfsigned-tls/main.go]: just out?

**Comment by klizhentas** [comment]
@colhom @xiang90 this PR seems really useful, do you know know when it is going to be merged? I'd be happy to help with it, just let me know.

File: pkg/util/etcdutil/member.go [hunk 18]
```diff
@@ -21,7 +21,7 @@ import (
 )
 
 type Member struct {
-	Name string
+	Name, Namespace string
```
Comment (hongchaodeng) [inline pkg/util/etcdutil/member.go]: We don't namespace var. We can assume etcd pods are within the same namespace

File: pkg/util/etcdutil/member.go [hunk 19]
```diff
@@ -38,15 +38,15 @@ func (m *Member) ClientAddr() string {
 		return strings.Join(m.ClientURLs, ",")
 	}
 
-	return fmt.Sprintf("http://%s:2379", m.Name)
+	return fmt.Sprintf("https://%s.%s.svc.cluster.local:2379", m.Name, m.Namespace)
```
Comment (hongchaodeng) [inline pkg/util/etcdutil/member.go]: You don't need to change this. We can assume they are within same ns.

File: pkg/spec/cluster.go [hunk 20]
```diff
@@ -99,6 +99,9 @@ type ClusterSpec struct {
 	// SelfHosted determines if the etcd cluster is used for a self-hosted
 	// Kubernetes cluster.
 	SelfHosted *SelfHostedPolicy `json:"selfHosted,omitempty"`
+
+	// etcd cluster TLS configuration
+	ClusterTLS *ClusterTLSPolicy `json:"clusterTLS"`
```
Comment (hongchaodeng) [inline pkg/spec/cluster.go]: Just `TLS` or `TLSSpec`.

File: pkg/spec/cluster_tls.go [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+package spec
+
+type ClusterTLSPolicy struct {
+	Static *StaticTLS `json:"static"`
+	//TODO: add more TLS options here
+}
+
+type StaticTLS struct {
+	// Secret containing peer and client CA cert files
+	CASecretName string `json:"caSecretName"`
```
Comment (hongchaodeng) [inline pkg/spec/cluster_tls.go]: `CASecret` without name. Same as others.

Comment (hongchaodeng) [inline pkg/spec/cluster_tls.go]: Why not combine client-CA into clientSecret and server-CA into serverSecret? This would make it more categorized.

**Comment by hongchaodeng** [comment]
Hi @colhom , We would like to make progress w.r.t. TLS feature. But a big PR like this would never work out fast and well. Can we close this PR and instead have multiple small, step-by-step PR?

I would suggest for the next step we should have a PR only for defining TLSSpec in ClusterSpec. Then we could implement based on the spec.

**Comment by colhom** [comment]
Breaking this up into pieces. First step is a PR to add tls spec [here](https://github.com/coreos/etcd-operator/pull/899/files)

**Pull request closed without merging**